fsevents: use shared FSEventStream

indutny opened the pull request:
It seems that number of simultaneously opened FSEventStreams is limited
on OSX (i.e. you can have only fixed number of them on one running
system), getting past through this limit will cause `FSEventStreamCreate` to
return false and write following message to stderr:

```
(CarbonCore.framework) FSEventStreamStart: register_with_server: ERROR:
f2d_register_rpc() => (null) (-21)
```

To prevent this, we must use only one shared FSEventStream with a paths for
all uv_fsevent_t handles, and then filter out events for each handle
using this paths again.

See https://github.com/joyent/node/issues/5463

indutny committed:
fsevents: use shared FSEventStream

It seems that number of simultaneously opened FSEventStreams is
limited on OSX (i.e. you can have only fixed number of them on
one running system), getting past through this limit will cause
`FSEventStreamCreate` to return false and write following message
to stderr:

    (CarbonCore.framework) FSEventStreamStart: register_with_server:
    ERROR: f2d_register_rpc() => (null) (-21)

To prevent this, we must use only one shared FSEventStream with a
paths for all uv_fsevent_t handles, and then filter out events for
each handle using this paths again.

See https://github.com/joyent/node/issues/5463

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit indutny/libuv@12852ba47cdb235e0276ee1afd3342e0946300fb has the following error(s):
- Commit message line too long: 4
- Commit message line too long: 7
- Commit message line too long: 10

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

indutny commented:
Should be ABI compatible with v0.10 if I hadn't messed with anything :)

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -72,26 +100,21 @@ static void uv__cf_loop_signal(uv_loop_t* loop,
 
 #define UV__FSEVENTS_WALK(handle, block)                                      \
```
While we're here, the name of this macro is somewhat poorly chosen.  'Walk' implies it just iterates but it also mutates it.  Can you address that in a follow-up commit?

indutny replied:
Sure, anything else to fix?

indutny committed:
fsevents: rename WALK => PROCESS

It isn't really walking through events, because it changes (resets)
them.

Per @bnoordhuis and @tjfontaine suggestion.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -125,42 +148,35 @@ static void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
+      path = paths[i];
+      len = strlen(path);
+
+      /* Filter out paths that are outside hanlde's request */
```
s/hanlde/handle/

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -125,42 +148,35 @@ static void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
+      len = strlen(path);
+
+      /* Filter out paths that are outside hanlde's request */
+      if (strstr(path, handle->realpath) != path)
```
Use strncmp() here. strstr() is potentially O(n*m).

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -169,79 +185,79 @@ static void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
+  /*
+   * NOTE: While we definitely could remember last event id,
+   * this will not be a right solution here. Because we might be watching
+   * different paths after reschedule.
```
Reschedule? I'm not sure I understand this comment.

indutny replied:
Well, imagine you're watching `/a` and last event id was `1`, then you're stopping it and starting watching `/b`. One might think (as I did) that its a good idea to start watching from event # `1`, but `/b` might already be at event-id `10000` and thus a lot of irrelevant data will be emitted.

bnoordhuis replied:
Ah, okay. Can you update the comment with that? :-)

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -381,11 +506,13 @@ static void* uv__cf_loop_runner(void* arg) {
+  state = loop->cf_state;
 
   uv_mutex_lock(&loop->cf_mutex);
   QUEUE_INIT(&split_head);
```
Apropos nothing, there is no reason for this to be inside the critical section.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -169,79 +185,79 @@ static void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
-        pos != path + 1)
-      continue;
+      /* Do not emit events from subdirectories (without option set) */
+      pos = strchr(path, '/');
```
Don't call strchr() when UV_FS_EVENT_RECURSIVE is not set.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -250,43 +266,113 @@ static void uv__fsevents_schedule(void* arg) {
+    if (paths[i] == NULL)
+      abort();
+  }
+  uv_mutex_unlock(&state->fsevent_mutex);
```
Does this critical section have to be this big? Can you move the allocation and the loop out of it somehow?

indutny replied:
Not really...

indutny replied:
Its not that big, actually... Consider it as we're doing this stuff in main thread.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -250,43 +266,113 @@ static void uv__fsevents_schedule(void* arg) {
 
   /* Notify main thread that we're done here */
-  uv_sem_post(&handle->cf_sem);
+  if ((handle->flags & (UV_CLOSING | UV_CLOSED)) != 0)
```
uv__is_closing?

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -70,28 +98,23 @@ static void uv__cf_loop_signal(uv_loop_t* loop,
-        event = QUEUE_DATA(curr, uv__fsevents_event_t, member);               \
-        QUEUE_REMOVE(curr);                                                   \
         /* Invoke block code, but only if handle wasn't closed */             \
         if (((handle)->flags & (UV_CLOSING | UV_CLOSED)) == 0)                \
```
uv__is_closing?

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -250,43 +266,113 @@ static void uv__fsevents_schedule(void* arg) {
+  state = malloc(sizeof(*state));
+  if (state == NULL)
+    return -ENOMEM;
+  memset(state, 0, sizeof(*state));
```
Just curious, why not calloc()?

indutny replied:
No reason, just added memset in separate pass. Will fix in follow-up commit.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -442,10 +572,13 @@ int uv__fsevents_init(uv_fs_event_t* handle) {
 
   /* Get absolute path to file */
   handle->realpath = realpath(handle->filename, NULL);
```
Calling realpath() with resolved_name=NULL is implementation-defined.  I'm reasonably sure older versions of OS X segfault when it's NULL.  (Though it's possible those don't support FSEvents in the first place.)

indutny replied:
It seems to be working, as this code was here for a half-year (or more).

bnoordhuis replied:
By the way, handle->realpath is leaked when the call to malloc() a few lines down fails.

bnoordhuis replied:
> It seems to be working, as this code was here for a half-year (or more).

"seems to be working" is not the same thing as "correct".  I guess it's possible there is simply no one still running OS X 10.5 except my dad-in-law.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -442,10 +572,13 @@ int uv__fsevents_init(uv_fs_event_t* handle) {
+    return -errno;
+  handle->realpath_len = strlen(handle->realpath);
+
+  /* Initialize signly-linked list */
```
Typo.  Hip startup name though.

bnoordhuis commented on src/unix/fsevents.c:
```diff
@@ -459,35 +592,54 @@ int uv__fsevents_init(uv_fs_event_t* handle) {
   handle->cf_cb->flags |= UV__HANDLE_INTERNAL;
   uv_unref((uv_handle_t*) handle->cf_cb);
 
+  /* TODO(indutny): check return value */
```
Well, do it then. :-)

indutny committed:
fixes

bnoordhuis commented:
Last round of comments. If you fix those then it LGTM.

indutny commented:
Thanks, landed with fixes in cd2794c01fc84a4118f79e31071cb1bca78918f5. Backporting to v0.10 now.

piscisaureus commented:
Nice work @indutny !

indutny commented:
Thanks :)

Waiting for merging backport before closing this issue: https://github.com/joyent/libuv/pull/896

siebeneicher commented:
Hey guys, I still get this error using grunt-watch on mac os. Nodejs is up to date (0.10.26). Does this fix practically fixed the mentioned error in the inital comment? indutny?
Short reply very much appreciated! THanks
